File: 빈출알고리즘/이항계수.py
# 10

file = open("C:/Users/onetu/PycharmProjects/Daily_Exercise_with_Programmers/백준/단계별/분할정복/이항계수3.txt", "r")

input = file.readline


# * 이항계수
# - 조합론 / 페르마의 소정리 / 모듈러 곱셈 역원 / 분할정복 / DP
# - 5^3 mod 3 = 5 mod 3
# - a^t mod t = a mod t
# - a^(t-2) mod t = a^-1 mod t


p = 10**9+7


# n! mod p
def factorial(n):
    result = 1
    while n > 1:
        result = (result * n) % p
        n -= 1
    return result


# a^(t-2) mod t -> (k!(n-k)!)^(p-2) mod p
def pow(a, tt):
    if tt <= 1:
        return a % p
    result = pow(a, tt//2)**2 % p
    if tt%2 != 0:
        result *= a % p
    return result


n, k = map(int, input().split())

numerator = factorial(n)
denominator = pow(factorial(k) * factorial(n-k), p-2)

print(numerator * denominator % p)


file.close()

File: src/빈출알고리즘/정수조합론/이항계수.java
package 빈출알고리즘.정수조합론;

import java.io.IOException;
import java.io.BufferedReader;
import java.io.FileReader;

public class 이항계수 {
    final static long P = 100000007L;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(
                new FileReader(
                        "C:\\Users\\onetu\\IdeaProjects\\Daily_Algo_Exercise\\백준\\단계별\\분할정복\\이항계수3.txt"
                )
        );

        String[] line = br.readLine().split(" ");

        int n = Integer.parseInt(line[0]);
        int k = Integer.parseInt(line[1]);

        long numerator = factorial(n);
        long denominater = pow(factorial(k) * factorial(n-k), P-2);

        System.out.println(numerator * denominater % P);
    }

    private static long factorial(int n) {
        long result = 1L;
        while (n > 1){
            result = (result * n) % P;
            n --;
        }
        return result;
    }

    private static long pow(long a, long b) {
        if (b <= 1)
            return a % P;

        long result = pow(a, b/2);
        if (b % 2 != 0)
            return (result * result % P) * (a % P);
        return result * result % P;
    }
}